tests/testsuite/build.rs

```diff
@@ -1563,6 +1579,372 @@ fn crate_env_vars() {
+                fn main() {
+                    // Verify CARGO_WORKSPACE_DIR isn't set for bins
+                    assert!(option_env!("CARGO_WORKSPACE_DIR").is_none());
+                }
+            "#,
+        )
+        .file(
+            "foo/src/lib.rs",
+            r#"
+                use std::env;
+
+                #[test]
+                fn env() {
+                    // Check that CARGO_WORKSPACE_DIR isn't set for unit tests
+                    assert!(option_env!("CARGO_WORKSPACE_DIR").is_none());
+                    env::var("CARGO_WORKSPACE_DIR").unwrap_err();
+                }
+            "#,
+        )
+        .file(
+            "foo/examples/ex-env-vars.rs",
+            r#"
+                fn main() {
+                    // Verify CARGO_WORKSPACE_DIR isn't set for examples
+                    assert!(option_env!("CARGO_WORKSPACE_DIR").is_none());
+                }
+            "#,
+        )
+        .file(
+            "foo/tests/env.rs",
+            r#"
+                use std::path::Path;
+
+                #[test]
+                fn env() {
+                    assert!(Path::new(option_env!("CARGO_WORKSPACE_DIR").unwrap()).join(file!()).exists());
+                }
+            "#,
+        )
+        .file(
+            "baz/Cargo.toml",
+            r#"
+            [workspace]
+            members = ["baz_member"]
+            "#,
+        )
+        .file(
+            "baz/baz_member/Cargo.toml",
+            r#"
+            [package]
+            name = "baz_member"
+            version = "0.1.0"
+            authors = []
+            "#,
+        )
+        .file("baz/baz_member/src/lib.rs", "#[test] fn baz_member() {}")
+        .file(
+            "baz/baz_member/tests/env.rs",
+            r#"
+            use std::path::Path;
+
+            #[test]
+            fn env() {
+                // is_primary = true
+                // is_workspace = true
+                // cargo_workspace_dir = "../.."
+                // cx.bcx.ws.root() = "/Users/yerke/dev/rust/cargo/target/tmp/cit/t0/foo"
+                // unit.pkg.root() = "/Users/yerke/dev/rust/cargo/target/tmp/cit/t0/foo/baz/baz_member"
+                // option_env!("CARGO_WORKSPACE_DIR").unwrap() = "../.."
+                // file!() = "baz/baz_member/tests/env.rs"
+                // std::env::current_dir().unwrap() = "/Users/yerke/dev/rust/cargo/target/tmp/cit/t0/foo/baz/baz_member"
+                // option_env!("CARGO_MANIFEST_DIR").unwrap() = "/Users/yerke/dev/rust/cargo/target/tmp/cit/t0/foo/baz/baz_member"
+                dbg!(option_env!("CARGO_WORKSPACE_DIR").unwrap());
+                dbg!(file!());
+                dbg!(std::env::current_dir().unwrap());
+                dbg!(option_env!("CARGO_MANIFEST_DIR").unwrap());
+                assert!(Path::new(option_env!("CARGO_WORKSPACE_DIR").unwrap()).join(file!()).exists());
+                // thread 'env' panicked at 'assertion failed: `(left == right)`
```

Comment:
This test fails. I added debug info above. Essentially, I would expect `CARGO_MANIFEST_DIR` to be `/Users/yerke/dev/rust/cargo/target/tmp/cit/t0/foo/baz`, but it's `/Users/yerke/dev/rust/cargo/target/tmp/cit/t0/foo` instead.
So in a situation where a member of workspace depends on path dependency crate, which is in another workspace and we test path dependency crate, we get the first workspace instead of the second one.

Reply:
Just to avoid any dusty corners of cargo, try moving `baz` to `/Users/yerke/dev/rust/cargo/target/tmp/cit/t0/foo/baz`.  I think the way to do this is with `project().at("baz')`.

My concern is that without `baz_member` having `package.workspace` or `workspace`, cargo is not seeing `baz/Cargo.toml` to realize `baz_member` is within a workspace.  Since `baz_member` is underneath `foo`, it may automatically become a member.  By splitting it out, this breaks the auto-workspace behavior.  You can go a step further by also setting `package.workspace`.

If that still doesn't work, have we verified that we are getting anything but the top-level workspace to work with?